### PR TITLE
[DSM] DDP-8465 | Fix for patch errors (onc_history_detail, onc_history, tissue, participant)

### PR DESCRIPTION
**Pull request by ScalaChamp**
## Context

Ticket: https://broadinstitute.atlassian.net/browse/DDP-8465

There were a lot of self-contained bugs in patch requests. The ticket describes just a single bug related to `onc_history_detail` only. The bugs required us to change the Patch instance from the backend side and for some cases change `primary_key` of `DBElement`-s too, I'll explain everything.

The problem description: A lot of patch requests use `dsm.participant.participantId` as a `Patch#parentId` to make patches in DB and ES. The issue is that sometimes participants don't have `dsm.participant` object in ES at all which causes `dsm.participant.participantId` to be undefined, thereby we can't make successful patches.

Solution: Me and @SimoneMaiwald decided to use `ddpParticipantId` as a `Patch#parentId` for the participants for whom we don't have `dsm.participant.participantId` on the client. As a result with the help of `ddpParticipantId` we can read `participantId` from DB because it's guaranteed to be there and act accordingly.

Technical Details:
- Create `PreProcessor<A>` generic interface which could span the series of general cases where we need to update the specific object based on custom logic
- Add `BasePatchPreProcessor` implementing `PreProcessor<Patch>` and describe one abstract method called `updatePatch()`
- Establish the static factory method `BasePatchPreProcessor.produce` which returns the sub-instances of `BasePatchPreProcessor` based on some logic
- Make `ParentRelatedPatchPreProcessor`, the special instance of `BasePatchPreProcessor` which changes the instance fields of patch, for more details look into `ParentRelatedPatchPreProcessor.java` line 46.
- Fix minor bugs (add `OncHistory.class` param in the static of block of `PatchUtil`, return `o.created` instead of `createdOncHistory` from `OncHistory.setOncHistoryCreated` function, add appropriate DB related annotations to `OncHistory`.
- Change the `PatchRoute.procesRequest` according to new design.
- Establish `OncHistoryDBElementBuider` for setting the correct primary key (`participant_id`) to the `DBElement` of `OncHistory`.
- Add a new instance of `ExistingRecordPatch` with the name of `ExistingOncHistoryPatch` for handling the patch of existing `OncHistory`. change the static factory method accordingly.
- Write unit tests covering all cases.

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/OncHistoryDBElementBuilder.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package org.broadinstitute.dsm.model.patch;
+
+import org.broadinstitute.dsm.db.structure.DBElement;
+import org.broadinstitute.dsm.statics.DBConstants;
+
+// The primary key of DBElement built by this class should equal to "participant_id".
```
Comment (devonbush) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/model/patch/OncHistoryDBElementBuilder.java:6]: this is a great comment, thanks!

**Review by devonbush — APPROVED** [review]
really appreciate the class comments

**Review by SimoneMaiwald — CHANGES_REQUESTED** [review]
Please add steps how to check the fix manually as well

**Comment by SimoneMaiwald** [comment]
fixed per https://github.com/broadinstitute/ddp-study-server/pull/2206

**Pull request closed without merging**